### PR TITLE
[ADDED] TokenHandler option for expiring token support

### PR DESCRIPTION
**Pull request by seriousben**
Aligning cnats with changes from https://github.com/nats-io/go-nats/pull/405

### CODE REVIEW

File: test/test.c [hunk 1]
```diff
@@ -7170,6 +7188,53 @@ test_AuthToken(void)
     _stopServer(serverPid);
 }
 
+static const char*
```
Comment (seriousben) [inline test/test.c]: Not testing the reconnect use case with TokenCb. Will do so if it feels required. (The test done in https://github.com/nats-io/go-nats/pull/405/files#diff-92c931046bdf2f3c164a4a28facea12fR152)

**Comment by seriousben** [comment]
Should now be aligned with https://github.com/nats-io/go-nats/pull/405.

File: src/opts.c [hunk 2]
```diff
@@ -194,6 +194,21 @@ natsOptions_SetToken(natsOptions *opts, const char *token)
...
 }
 
+natsStatus
+natsOptions_SetTokenHandler(natsOptions *opts, natsTokenHandler natsTokenCb, void *closure)
+{
+    natsStatus  s = NATS_OK;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    opts->tokenCb= natsTokenCb;
```
[the remark this answers is not in the document]
Comment (seriousben) [reply src/opts.c]: I am too used to gofmt these days :)

**Comment by kozlovic** [comment]
@seriousben @janmiderback Should we try to make it behave like the Go client in that it should fail if both token and token handlers are set?

**Comment by seriousben** [comment]
@kozlovic Yes. I am almost done. I am also tackling the reconnect test, but it is a bit more challenging. I'll push some progress soon.

File: src/status.h [hunk 3]
```diff
@@ -85,6 +85,8 @@ typedef enum
                                         ///  the subscription has previously been closed.
     NATS_INVALID_TIMEOUT,               ///< Timeout must be positive numbers.
 
+    NATS_TOKEN_ALREADY_SET,             ///< Token and token handler options cannot be both set.
```
[the remark this answers is not in the document]
Comment (kozlovic) [reply src/status.h]: After thinking about it, I think we should just use an existing status and set a specific error string. See the comments in places where you set the error.

File: src/conn.c [hunk 4]
```diff
@@ -766,6 +766,15 @@ _connectProto(natsConnection *nc, char **proto)
         pwd   = opts->password;
         token = opts->token;
     }
+
+    if (opts->tokenCb != NULL)
+    {
+        if (token != NULL)
+            return NATS_TOKEN_ALREADY_SET;
```
Comment (kozlovic) [inline src/conn.c]: Would replace with:
```
return nats_setError(NATS_ILLEGAL_STATE, "%s", "Token and token handler options cannot be both set");
```

File: src/opts.c [hunk 5]
```diff
@@ -180,13 +180,40 @@ natsOptions_SetToken(natsOptions *opts, const char *token)
 
     LOCK_AND_CHECK_OPTIONS(opts, 0);
 
-    NATS_FREE(opts->token);
-    opts->token= NULL;
-    if (token != NULL)
+    if ((token != NULL) && (opts->tokenCb != NULL)) {
+        s = nats_setDefaultError(NATS_TOKEN_ALREADY_SET);
```
Comment (kozlovic) [inline src/opts.c]: Would replace with:
```
s = nats_setError(NATS_ILLEGAL_STATE, "%s", "Cannot set a token if a token handler has already been set");
```

File: src/opts.c [hunk 6]
```diff
@@ -180,13 +180,40 @@ natsOptions_SetToken(natsOptions *opts, const char *token)
...
+
+natsStatus
+natsOptions_SetTokenHandler(natsOptions *opts, natsTokenHandler tokenCb, void *closure)
+{
+    natsStatus  s = NATS_OK;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    if ((tokenCb != NULL) && (opts->token != NULL)) {
+        s = nats_setDefaultError(NATS_TOKEN_ALREADY_SET);
```
Comment (kozlovic) [inline src/opts.c]: Would replace with:
```
s = nats_setError(NATS_ILLEGAL_STATE, "%s", "Cannot set a token handler if a token has already been set");
```

File: test/test.c [hunk 7]
```diff
@@ -7170,6 +7189,67 @@ test_AuthToken(void)
...
+    natsConnection_Destroy(nc);
+    nc = NULL;
+
+    test("cannot set a tokenHandler when token set: ");
+    s = natsOptions_SetTokenHandler(opts, NULL, NULL);
+    if (s == NATS_OK)
+        s = natsOptions_SetToken(opts, "token");
+    if (s == NATS_OK)
+        s = natsOptions_SetTokenHandler(opts, _tokenHandler, (char*) "testSecret");
+    testCond(s == NATS_TOKEN_ALREADY_SET);
```
Comment (kozlovic) [inline test/test.c]: Would replace those with NATS_ILLEGAL_STATE. You could (but don't have to), check that the last error we get contains the string that you expect. Again, we don't have to go there.

File: src/conn.c (hunk 4 above)
Comment (seriousben) [reply src/conn.c]: done

File: src/status.c [hunk 8]
```diff
@@ -39,6 +39,8 @@ static const char *statusText[] = {
     "Invalid Subscription",
     "Invalid Timeout",
 
+    "Token and tokenHandler options cannot be both set",
```
[the remark this answers is not in the document]
Comment (seriousben) [reply src/status.c]: removed as requested.

File: src/nats.h [hunk 9]
```diff
@@ -273,6 +273,15 @@ typedef natsStatus (*natsEvLoop_WriteAddRemove)(
 typedef natsStatus (*natsEvLoop_Detach)(
         void            *userData);
 
+/** \brief Callback used to build a token on connections and reconnections.
+  *
```
[the remark this answers is not in the document]
Comment (seriousben) [reply src/nats.h]: Good find. Thanks.

File: src/opts.c [hunk 10]
```diff
@@ -180,13 +180,40 @@ natsOptions_SetToken(natsOptions *opts, const char *token)
 
     LOCK_AND_CHECK_OPTIONS(opts, 0);
 
-    NATS_FREE(opts->token);
-    opts->token= NULL;
-    if (token != NULL)
+    if ((token != NULL) && (opts->tokenCb != NULL)) {
```
[the remark this answers is not in the document]
Comment (seriousben) [reply src/opts.c]: done

File: src/opts.c [hunk 11]
```diff
@@ -180,13 +180,40 @@ natsOptions_SetToken(natsOptions *opts, const char *token)
...
+}
+
+natsStatus
+natsOptions_SetTokenHandler(natsOptions *opts, natsTokenHandler tokenCb, void *closure)
+{
+    natsStatus  s = NATS_OK;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    if ((tokenCb != NULL) && (opts->token != NULL)) {
```
[the remark this answers is not in the document]
Comment (seriousben) [reply src/opts.c]: done

File: src/opts.c (hunk 5 above)
Comment (seriousben) [reply src/opts.c]: done

File: src/opts.c (hunk 6 above)
Comment (seriousben) [reply src/opts.c]: done

File: test/test.c [hunk 12]
```diff
@@ -7170,6 +7189,67 @@ test_AuthToken(void)
...
+{
+    natsStatus          s;
+    natsConnection      *nc       = NULL;
+    natsPid             serverPid = NATS_INVALID_PID;
+    natsOptions         *opts     = NULL;
+
+    serverPid = _startServer("nats://127.0.0.1:8232", "-auth testSecret -p 8232", false);
+    CHECK_SERVER_STARTED(serverPid);
+
+    nats_Sleep(1000);
```
[the remark this answers is not in the document]
Comment (seriousben) [reply test/test.c]: done.

File: test/test.c [hunk 13]
```diff
@@ -10824,6 +10904,122 @@ test_BasicClusterReconnect(void)
...
+static const char*
+_reconnectTokenHandler(void* closure)
+{
+    const char *token;
+    struct threadArg *args      = (struct threadArg*) closure;
+
+    natsMutex_Lock(args->m);
+    token = args->tokens[args->tokenCallCount % (sizeof(args->tokens)/sizeof(char*))];
+    args->tokenCallCount++;
+    printf("\ntoken %s\n", token);
```
[the remark this answers is not in the document]
Comment (seriousben) [reply test/test.c]: thanks. Done.

File: test/test.c [hunk 14]
```diff
@@ -10824,6 +10904,122 @@ test_BasicClusterReconnect(void)
...
+        s = natsOptions_SetReconnectedCB(opts, _reconnectedCb, (void*) &args);
+    if (s == NATS_OK)
+        s = natsOptions_SetMaxReconnect(opts, 10);
+    if (s == NATS_OK)
+        s = natsOptions_SetReconnectWait(opts, 100);
+
+    if (s != NATS_OK)
+        FAIL("Unable to setup test");
+
+    serverPid1 = _startServer("nats://127.0.0.1:22222", "-p 22222 --auth token1", false);
```
[the remark this answers is not in the document]
Comment (seriousben) [reply test/test.c]: Great idea. Done

File: test/test.c [hunk 15]
```diff
@@ -10824,6 +10904,122 @@ test_BasicClusterReconnect(void)
...
+
+    serverPid3 = _startServer("nats://127.0.0.1:22224", "-p 22224 --auth token3", false);
+    if (serverPid3 == NATS_INVALID_PID)
+    {
+        _stopServer(serverPid1);
+        _stopServer(serverPid2);
+    }
+    CHECK_SERVER_STARTED(serverPid3);
+
+    nats_Sleep(1000);
```
[the remark this answers is not in the document]
Comment (seriousben) [reply test/test.c]: done

**Comment by seriousben** [comment]
@kozlovic I think I have handled all of your comments. Thanks a lot.

**Comment by kozlovic** [comment]
@seriousben Thanks! I may not be able to get to it right away, but appreciate your contribution. I will review/merge as soon as I can.

**Comment by kozlovic** [comment]
@seriousben @janmiderback Sorry for the delay and thank you guys for the contribution!

**Pull request merged**